features/lorawan/lorastack/mac/LoRaMac.h

```diff
@@ -0,0 +1,353 @@
+ *              (______/|_____)_|_|_| \__)_____)\____)_| |_|
+ *              (C)2013 Semtech
+ *
+ *               ___ _____ _   ___ _  _____ ___  ___  ___ ___
+ *              / __|_   _/_\ / __| |/ / __/ _ \| _ \/ __| __|
+ *              \__ \ | |/ _ \ (__| ' <| _| (_) |   / (__| _|
+ *              |___/ |_/_/ \_\___|_|\_\_| \___/|_|_\\___|___|
+ *              embedded.connectivity.solutions===============
+ *
+ * \endcode
+ *
+ * \author    Miguel Luis ( Semtech )
+ *
+ * \author    Gregory Cristian ( Semtech )
+ *
+ * \author    Daniel Jaeckle ( STACKFORCE )
+ *
+ * \defgroup  LORAMAC LoRa MAC layer implementation
+ *            This module specifies the API implementation of the LoRaMAC layer.
+ *            This is a placeholder for a detailed description of the LoRaMac
+ *            layer and the supported features.
+ *
+ * Copyright (c) 2017, Arm Limited and affiliates.
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ */
+#ifndef __LORAMAC_H__
+#define __LORAMAC_H__
+
+#include "lorawan/system/LoRaWANTimer.h"
+#include "netsocket/LoRaRadio.h"
+#include "lorastack/phy/LoRaPHY.h"
+#include "lorawan/system/lorawan_data_structures.h"
+
+/*!
+ * Check the MAC layer state every MAC_STATE_CHECK_TIMEOUT in ms.
+ */
+#define MAC_STATE_CHECK_TIMEOUT                     1000
+
+/*!
+ * The maximum number of times the MAC layer tries to get an acknowledge.
+ */
+#define MAX_ACK_RETRIES                             8
+
+/*!
+ * The frame direction definition for uplink communications.
+ */
+#define UP_LINK                                     0
+
+/*!
+ * The frame direction definition for downlink communications.
+ */
+#define DOWN_LINK                                   1
+
+
+/*!
+ * LoRaMAC max EIRP (dBm) table.
+ */
+static const uint8_t LoRaMacMaxEirpTable[] = { 8, 10, 12, 13, 14, 16, 18, 20, 21, 24, 26, 27, 29, 30, 33, 36 };
+
+
+/*!
+ * \brief   LoRaMAC layer initialization
+ *
+ * \details In addition to the initialization of the LoRaMAC layer, this
+ *          function initializes the callback primitives of the MCPS and
+ *          MLME services. Every data field of \ref LoRaMacPrimitives_t must be
+ *          set to a valid callback function.
+ *
+ * \param   primitives [in] - A pointer to the structure defining the LoRaMAC
+ *                            event functions. Refer to \ref LoRaMacPrimitives_t.
+ *
+ * \param   callbacks [in] - A pointer to the structure defining the LoRaMAC
+ *                        callback functions. Refer to \ref LoRaMacCallback_t.
+ *
+ * \param   phy [in]- A reference to the selected PHY layer.
+ *
+ * \retval  `LoRaMacStatus_t` The status of the operation. The possible values are:
+ *          \ref LORAMAC_STATUS_OK
+ *          \ref LORAMAC_STATUS_PARAMETER_INVALID
+ *          \ref LORAMAC_STATUS_REGION_NOT_SUPPORTED.
+ */
+LoRaMacStatus_t LoRaMacInitialization( LoRaMacPrimitives_t *primitives, LoRaMacCallback_t *callbacks, LoRaPHY& phy);
```

Comment:
the above files (like LoRaWANStack) look different, this file is suddenly using camelcase in names, functions? I would assume this should be `loramac_status_t loramac_initialization(loramac_primitives_t ... )` ?

Reply:
We can't change these. This file is coming from semtech reference design. Until unless we totally ditch their code and write our own, I am not in favour of changing it.

Reply:
These files are treated as SDK from a partner?

Reply:
Anything below LoRaWANInterface is mixed code. At least at Mac level we have clear demarcation, i.e., LoRaMac.cpp is predominantly partner's code., except the introduction of a private class which runs a thread 
All implementations of LoRaPHY class (found in lorwan/lorastack/phy/) are using partner's code heavily alongwith our design. LoRaMacCrypto has APIs which were defined by partner but the implementation inside is MbedTLS based. It's hard to isolate bits and pieces. May be we can go through this together on skype ?